File: nodejs/ping-pong/ponger.js
const types = require('@lokomotes/station-types')
const MsgType = types.MsgType

// allow only one station to link.
/**
 * @param {types.Station} station
 * @param {Array<string>} args
 */
module.exports = types.template.accept(1, async (station, msg, src) => {
  let pong = (dst) => {
    station.log('pong')
    setTimeout(() => {
      // send signal to `dst` with message "ponged".
      station.send(MsgType.Signal, 'ponged').to(dst)
    }, 1000)
  }

  // listen signal from linked station.
  station.grab(src).on(MsgType.Signal, (msg) => {
    station.log(`${msg} from ${src.name}`)
    pong(src)
  })

  // listen block message from linked station.
  station.grab(src).on(MsgType.Block, () => {
    // the `ponger` can not send messages to the linked Station anymore,
    // so there is nothing to do.

    // stop this activity.
    station.close()
  })
})

File: dotnet/ping-pong/Ponger.cs
using System;
using System.Threading.Tasks;
using Loko.Station;

namespace PingPong
{
    public class Ponger : Loko.Station.Template.Accept
    {
        private StationDesc pinger;

        public Ponger() : base(1) { }

        public override void Start(string message, StationDesc src)
        {
            pinger = src;

            var grabPinger = Station.Grab(pinger);
            grabPinger.Signaled += handlePing;
            grabPinger.Blocked += handleBlock;
        }

        private async void pong()
        {
            await Task.Delay(1000);
            Station.Log("pong");
            await Station.Send(MsgType.Signal, "ponged").To(pinger);
        }

        private void handlePing(string msg, StationDesc src)
        {
            Station.Log($"{msg} from {src.Name}");
            pong();
        }

        private void handleBlock(string msg, StationDesc src)
        {
            Station.Close();
        }
    }
}